File: proxies/stubs/java/src/com/chaosgroup/vraycloud/stubs/GeneratedTest.java
/**
 * Test API 1.0
 *
 * Part of the JSON-WS library - Java 1.7 Proxy
 * Copyright (c) 2013-2014 ChaosGroup. All rights reserved.
 *
 * This code uses the following libraries
 * 	- com.google.code.gson:gson:2.2.4
 *	- org.java-websocket:Java-WebSocket:1.3.0
 */

package com.chaosgroup.jsonws.stubs;

import java.net.MalformedURLException;
import java.net.URISyntaxException;
import java.net.URL;
import com.google.gson.*;

/**
 * Test API 1.0 Proxy
 */
public class GeneratedTest implements AutoCloseable {

	// RPC tunnel - all method calls are piped here
	private RpcTunnel rpcTunnel;
	// The transport mechanism used by the tunnel for each method call
	private RpcTunnel.Transport defaultTransport = RpcTunnel.Transport.HTTP;

	/**
	 * Constructs a new proxy using the specified URL
	 * @param url Full URL of the web service endpoint.
	 * @throws java.net.MalformedURLException
	 * @throws java.net.URISyntaxException
	 */
	public GeneratedTest(String url) throws MalformedURLException, URISyntaxException {
		rpcTunnel = new RpcTunnel(url, rpcEventHandler);
	}

	/**
	 * Sets the default transport mechanism to HTTP
	 */
	public GeneratedTest useHTTP() {
		defaultTransport = RpcTunnel.Transport.HTTP;
		return this;
	}

	/**
	 * Sets the default transport mechanism to WebSocket
	 */
	public GeneratedTest useWS() {
		defaultTransport = RpcTunnel.Transport.WebSocket;
		return this;
	}

	@Override
	public void close() throws Exception {
		rpcTunnel.close();
	}

	private TestEventHandler testEventHandler;
	public interface TestEventHandler {
		public static final String Name = "testEvent";
		void onTestEvent(Long data);
	}
	public void onTestEvent(TestEventHandler eventHandler) {
		testEventHandler = eventHandler;
		rpcTunnel.call(testEventHandler == null ? "rpc.off" : "rpc.on",
				new Object[] { TestEventHandler.Name }, false, RpcTunnel.Transport.WebSocket);
	}

	private TestEvent2Handler testEvent2Handler;
	public interface TestEvent2Handler {
		public static final String Name = "testEvent2";
		void onTestEvent2(RenderOptions[] data);
	}
	public void onTestEvent2(TestEvent2Handler eventHandler) {
		testEvent2Handler = eventHandler;
		rpcTunnel.call(testEvent2Handler == null ? "rpc.off" : "rpc.on",
				new Object[] { TestEvent2Handler.Name }, false, RpcTunnel.Transport.WebSocket);
	}

	private TestEvent3Handler testEvent3Handler;
	public interface TestEvent3Handler {
		public static final String Name = "testEvent3";
		void onTestEvent3(JsonObject data);
	}
	public void onTestEvent3(TestEvent3Handler eventHandler) {
		testEvent3Handler = eventHandler;
		rpcTunnel.call(testEvent3Handler == null ? "rpc.off" : "rpc.on",
				new Object[] { TestEvent3Handler.Name }, false, RpcTunnel.Transport.WebSocket);
	}

	private TestEvent4Handler testEvent4Handler;
	public interface TestEvent4Handler {
		public static final String Name = "testEvent4";
		void onTestEvent4(Boolean data);
	}
	public void onTestEvent4(TestEvent4Handler eventHandler) {
		testEvent4Handler = eventHandler;
		rpcTunnel.call(testEvent4Handler == null ? "rpc.off" : "rpc.on",
				new Object[] { TestEvent4Handler.Name }, false, RpcTunnel.Transport.WebSocket);
	}

	private TestBinaryEventHandler testBinaryEventHandler;
	public interface TestBinaryEventHandler {
		public static final String Name = "testBinaryEvent";
		void onTestBinaryEvent(byte[] data);
	}
	public void onTestBinaryEvent(TestBinaryEventHandler eventHandler) {
		testBinaryEventHandler = eventHandler;
		rpcTunnel.call(testBinaryEventHandler == null ? "rpc.off" : "rpc.on",
				new Object[] { TestBinaryEventHandler.Name }, false, RpcTunnel.Transport.WebSocket);
	}

	private Ns1_testEvent1Handler ns1_testEvent1Handler;
	public interface Ns1_testEvent1Handler {
		public static final String Name = "ns1.testEvent1";
		void onNs1_testEvent1();
	}
	public void onNs1_testEvent1(Ns1_testEvent1Handler eventHandler) {
		ns1_testEvent1Handler = eventHandler;
		rpcTunnel.call(ns1_testEvent1Handler == null ? "rpc.off" : "rpc.on",
				new Object[] { Ns1_testEvent1Handler.Name }, false, RpcTunnel.Transport.WebSocket);
	}


	@SuppressWarnings("FieldCanBeLocal")
	private final RpcTunnel.EventHandler rpcEventHandler = new RpcTunnel.EventHandler() {
		@Override
		public void onEvent(String eventId, JsonElement eventData) {
			switch (eventId) {

				case TestEventHandler.Name:
					if (testEventHandler != null) {
						testEventHandler.onTestEvent((Long) new Gson().fromJson(eventData, Long.class));
					}
					break;

				case TestEvent2Handler.Name:
					if (testEvent2Handler != null) {
						testEvent2Handler.onTestEvent2((RenderOptions[]) new Gson().fromJson(eventData, RenderOptions[].class));
					}
					break;

				case TestEvent3Handler.Name:
					if (testEvent3Handler != null) {
						testEvent3Handler.onTestEvent3(eventData.getAsJsonObject());
					}
					break;

				case TestEvent4Handler.Name:
					if (testEvent4Handler != null) {
						testEvent4Handler.onTestEvent4(eventData.getAsBoolean());
					}
					break;

				case TestBinaryEventHandler.Name:
					if (testBinaryEventHandler != null) {
						testBinaryEventHandler.onTestBinaryEvent(javax.xml.bind.DatatypeConverter.parseBase64Binary(eventData.getAsString()));
					}
					break;

				case Ns1_testEvent1Handler.Name:
					if (ns1_testEvent1Handler != null) {
						ns1_testEvent1Handler.onNs1_testEvent1();
					}
					break;

			}
		}
	};


	public static class BaseRpcObject {
		@Override
		public String toString() {
			return new Gson().toJson(this);
		}
	}

	/**
	 *
	 */
	public static enum RenderMode {
		Production, RtCpu, RtGpuCuda
	}

	/**
	 *
	 */
	public static class RenderOptions extends BaseRpcObject {
		//
		public Long width;
		//
		public Long height;
		//
		public RenderMode renderMode;
	}

	/**
	 *
	 */
	public static class DefaultArray extends BaseRpcObject {
		//
		public String[] property;
	}




	public final class ns1 {
		public final class sub1 {
			public final class sub2 {
				/**
				 *
				 */
				public void method1() {
					rpcTunnel.call("ns1.sub1.sub2.method1", new Object[] {  }, false, defaultTransport);
				}

			}
			public final sub2 sub2 = new sub2();

		}
		public final sub1 sub1 = new sub1();

		/**
		 *
		 */
		public ProxyFuture<String> method1() {
			return new ProxyFuture<String>(rpcTunnel.call("ns1.method1",
					new Object[] {  }, true, defaultTransport)) {
				@Override
				protected String convert(JsonElement result) {
					return result.getAsString();
				}
			};
		}

	}
	public final ns1 ns1 = new ns1();

	public final class ns2 {
		public final class sub1 {
			public final class sub2 {
				/**
				 *
				 */
				public void method1() {
					rpcTunnel.call("ns2.sub1.sub2.method1", new Object[] {  }, false, defaultTransport);
				}

			}
			public final sub2 sub2 = new sub2();

		}
		public final sub1 sub1 = new sub1();

	}
	public final ns2 ns2 = new ns2();

	/**
	 * Some test method example, does int sum
	 * @param a
	 * @param b
	 */
	public ProxyFuture<Long> sum(Long a, Long b) {
		return new ProxyFuture<Long>(rpcTunnel.call("sum",
				new Object[] { a, b }, true, defaultTransport)) {
			@Override
			protected Long convert(JsonElement result) {
				return (Long) new Gson().fromJson(result, Long.class);
			}
		};
	}

	/**
	 *
	 */
	public void sumReturn() {
		rpcTunnel.call("sumReturn", new Object[] {  }, false, defaultTransport);
	}

	/**
	 *
	 * @param a
	 */
	public ProxyFuture<RenderOptions> echo(RenderOptions a) {
		return new ProxyFuture<RenderOptions>(rpcTunnel.call("echo",
				new Object[] { a }, true, defaultTransport)) {
			@Override
			protected RenderOptions convert(JsonElement result) {
				return (RenderOptions) new Gson().fromJson(result, RenderOptions.class);
			}
		};
	}

	/**
	 *
	 * @param a
	 */
	public ProxyFuture<JsonObject> echoObject(JsonElement a) {
		return new ProxyFuture<JsonObject>(rpcTunnel.call("echoObject",
				new Object[] { a }, true, defaultTransport)) {
			@Override
			protected JsonObject convert(JsonElement result) {
				return result.getAsJsonObject();
			}
		};
	}

	/**
	 *
	 */
	public ProxyFuture<Long> throwError() {
		return new ProxyFuture<Long>(rpcTunnel.call("throwError",
				new Object[] {  }, true, defaultTransport)) {
			@Override
			protected Long convert(JsonElement result) {
				return (Long) new Gson().fromJson(result, Long.class);
			}
		};
	}

	/**
	 *
	 */
	public void testMe() {
		rpcTunnel.call("testMe", new Object[] {  }, false, defaultTransport);
	}

	/**
	 *
	 */
	public ProxyFuture<Void> testMe1() {
		return new ProxyFuture<Void>(rpcTunnel.call("testMe1",
				new Object[] {  }, true, defaultTransport)) {
			@Override
			protected Void convert(JsonElement result) {
				return null;
			}
		};
	}

	/**
	 * A sample method.
	 * @param a A simple string parameter.
	 */
	public ProxyFuture<String> testMe2(String a) {
		return new ProxyFuture<String>(rpcTunnel.call("testMe2",
				new Object[] { a }, true, defaultTransport)) {
			@Override
			protected String convert(JsonElement result) {
				return result.getAsString();
			}
		};
	}

	/**
	 *
	 */
	public void testMe3() {
		rpcTunnel.call("testMe3", new Object[] {  }, false, defaultTransport);
	}

	/**
	 *
	 */
	public void testMe4() {
		rpcTunnel.call("testMe4", new Object[] {  }, false, defaultTransport);
	}

	/**
	 *
	 * @param p
	 */
	public void TestDefaultArray(DefaultArray p) {
		rpcTunnel.call("TestDefaultArray", new Object[] { p }, false, defaultTransport);
	}

	/**
	 *
	 * @param u
	 */
	public ProxyFuture<URL> TestUrl(URL u) {
		return new ProxyFuture<URL>(rpcTunnel.call("TestUrl",
				new Object[] { u }, true, defaultTransport)) {
			@Override
			protected URL convert(JsonElement result) throws MalformedURLException {
				return new URL(result.getAsString());
			}
		};
	}

	/**
	 *
	 */
	public ProxyFuture<RenderOptions[]> getRenderOptions() {
		return new ProxyFuture<RenderOptions[]>(rpcTunnel.call("getRenderOptions",
				new Object[] {  }, true, defaultTransport)) {
			@Override
			protected RenderOptions[] convert(JsonElement result) {
				return (RenderOptions[]) new Gson().fromJson(result, RenderOptions[].class);
			}
		};
	}

	/**
	 *
	 * @param theString
	 */
	public ProxyFuture<byte[]> echoStringAsBuffer(String theString) {
		return new ProxyFuture<byte[]>(rpcTunnel.call("echoStringAsBuffer",
				new Object[] { theString }, true, defaultTransport)) {
			@Override
			protected byte[] convert(JsonElement result) {
				return javax.xml.bind.DatatypeConverter.parseBase64Binary(result.getAsString());
			}
		};
	}

	/**
	 *
	 * @param buffer
	 */
	public ProxyFuture<Long> getBufferSize(byte[] buffer) {
		return new ProxyFuture<Long>(rpcTunnel.call("getBufferSize",
				new Object[] { javax.xml.bind.DatatypeConverter.printBase64Binary(buffer) }, true, defaultTransport)) {
			@Override
			protected Long convert(JsonElement result) {
				return (Long) new Gson().fromJson(result, Long.class);
			}
		};
	}

	/**
	 *
	 * @param n
	 */
	public ProxyFuture<Long[]> returnFrom0ToN(Long n) {
		return new ProxyFuture<Long[]>(rpcTunnel.call("returnFrom0ToN",
				new Object[] { n }, true, defaultTransport)) {
			@Override
			protected Long[] convert(JsonElement result) {
				return (Long[]) new Gson().fromJson(result, Long[].class);
			}
		};
	}

	/**
	 *
	 * @param required
	 */
	public void optionalArgs(Boolean required) {
		rpcTunnel.call("optionalArgs", new Object[] { required }, false, defaultTransport);
	}

	/**
	 *
	 * @param required
	 * @param p1
	 */
	public void optionalArgs(Boolean required, Long p1) {
		rpcTunnel.call("optionalArgs", new Object[] { required, p1 }, false, defaultTransport);
	}

	/**
	 *
	 * @param required
	 * @param p1
	 * @param p2
	 */
	public void optionalArgs(Boolean required, Long p1, Long p2) {
		rpcTunnel.call("optionalArgs", new Object[] { required, p1, p2 }, false, defaultTransport);
	}

	/**
	 *
	 * @param ints
	 */
	public ProxyFuture<Long> sumArray(Long[] ints) {
		return new ProxyFuture<Long>(rpcTunnel.call("sumArray",
				new Object[] { ints }, true, defaultTransport)) {
			@Override
			protected Long convert(JsonElement result) {
				return (Long) new Gson().fromJson(result, Long.class);
			}
		};
	}

	/**
	 *
	 * @param a
	 */
	public ProxyFuture<JsonElement> testAny(JsonElement a) {
		return new ProxyFuture<JsonElement>(rpcTunnel.call("testAny",
				new Object[] { a }, true, defaultTransport)) {
			@Override
			protected JsonElement convert(JsonElement result) {
				return result;
			}
		};
	}

}

File: test/generated-test.js
(function (exports) {
	var RpcClient = (function () {
		try {
			return require('json-ws/client');
		} catch (e) {
			return exports.RpcClient;
		}
	}());
	var inherits = RpcClient.ports.inherits;
	var EventEmitter = RpcClient.ports.EventEmitter;
	var RpcTunnel = RpcClient.tunnel;

	var GeneratedTest = exports.GeneratedTest = function GeneratedTest(url, sslSettings) {
		if (!this instanceof GeneratedTest) {
			return new GeneratedTest(url);
		}
		if (!url || typeof url !== 'string') {
			throw new Error('Invalid proxy URL');
		}
		var self = this;
		this.defaultTransport = 'http';
		this.rpc = new RpcTunnel(url, sslSettings);
		this.rpc.on('event', function(e) {
			self.emit(e.name, e.data);
		});
		function rebind(obj) {
			var result = {};
			for (var i in obj) {
				var prop = obj[i];
				if (typeof prop === 'function') {
					result[i] = prop.bind(self);
				} else if (typeof prop === 'object') {
					result[i] = rebind(prop);
				}
			}
			return result;
		}
		for (var i in this) {
			if (this[i] && this[i]._ns) {
				this[i] = rebind(this[i]);
			}
		}
	};
	inherits(GeneratedTest, EventEmitter);

	GeneratedTest.prototype.useHTTP = function() {
		this.defaultTransport = 'http';
		return this;
	};

	GeneratedTest.prototype.useWS = function() {
		this.defaultTransport = 'ws';
		return this;
	};

	GeneratedTest.prototype.close = function() {
		this.rpc.close();
	};

	GeneratedTest.prototype.on = GeneratedTest.prototype.addListener = function(type, listener) {
		if (this.listeners(type).length == 0) {
			this.rpc.call({ method: 'rpc.on', params: [type], transport: 'ws' });
		}
		EventEmitter.prototype.addListener.call(this, type, listener);
	};

	GeneratedTest.prototype.removeListener = function(type, listener) {
		EventEmitter.prototype.removeListener.call(this, type, listener);
		if (this.listeners(type).length == 0) {
			this.rpc.call({ method: 'rpc.off', params: [type], transport: 'ws' });
		}
	};

	GeneratedTest.prototype.removeAllListeners = function(type) {
		EventEmitter.prototype.removeAllListeners.call(this, type);
		this.rpc.call({ method: 'rpc.off', params: [type], transport: 'ws' });
	};

	GeneratedTest.RenderMode = function (val) {
		switch (val) {
		case 'Production': return -1;
		case 'RtCpu': return 0;
		case 'RtGpuCuda': return 5;
		case -1: return 'Production';
		case 0: return 'RtCpu';
		case 5: return 'RtGpuCuda';
		}
	};
	GeneratedTest.RenderMode.Production = -1;
	GeneratedTest.RenderMode.RtCpu = 0;
	GeneratedTest.RenderMode.RtGpuCuda = 5
	Object.freeze(GeneratedTest.RenderMode);

	GeneratedTest.prototype.ns1 = {_ns:true};
	GeneratedTest.prototype.ns1.sub1 = {_ns:true};
	GeneratedTest.prototype.ns1.sub1.sub2 = {_ns:true};
	GeneratedTest.prototype.ns2 = {_ns:true};
	GeneratedTest.prototype.ns2.sub1 = {_ns:true};
	GeneratedTest.prototype.ns2.sub1.sub2 = {_ns:true};

	/**
	 * Some test method example, does int sum
	 * @param {number} a 
	 * @param {number} b 
	 * @returns {number}
	 */
	GeneratedTest.prototype.sum = function(a, b) {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = Math.min(2, args.length);
		return this.rpc.call({
			method: 'sum',
			params: args,
			expectReturn: true,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * 
	 */
	GeneratedTest.prototype.sumReturn = function() {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = 0;
		return this.rpc.call({
			method: 'sumReturn',
			params: args,
			expectReturn: false,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * 
	 * @param {RenderOptions} a 
	 * @returns {RenderOptions}
	 */
	GeneratedTest.prototype.echo = function(a) {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = Math.min(1, args.length);
		return this.rpc.call({
			method: 'echo',
			params: args,
			expectReturn: true,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * 
	 * @param {object} a 
	 * @returns {object}
	 */
	GeneratedTest.prototype.echoObject = function(a) {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = Math.min(1, args.length);
		return this.rpc.call({
			method: 'echoObject',
			params: args,
			expectReturn: true,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * 
	 * @returns {number}
	 */
	GeneratedTest.prototype.throwError = function() {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = 0;
		return this.rpc.call({
			method: 'throwError',
			params: args,
			expectReturn: true,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * 
	 * @returns {object[]}
	 */
	GeneratedTest.prototype.throwUnexpectedError = function() {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = 0;
		return this.rpc.call({
			method: 'throwUnexpectedError',
			params: args,
			expectReturn: true,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * A sample method.
	 * @returns {object}
	 */
	GeneratedTest.prototype.testMe = function() {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = 0;
		return this.rpc.call({
			method: 'testMe',
			params: args,
			expectReturn: true,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * 
	 */
	GeneratedTest.prototype.testMe1 = function() {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = 0;
		return this.rpc.call({
			method: 'testMe1',
			params: args,
			expectReturn: true,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * A sample method.
	 * @param {string} a A simple string parameter.
	 * @returns {string}
	 */
	GeneratedTest.prototype.testMe2 = function(a) {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = Math.min(1, args.length);
		return this.rpc.call({
			method: 'testMe2',
			params: args,
			expectReturn: true,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * 
	 */
	GeneratedTest.prototype.testMe3 = function() {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = 0;
		return this.rpc.call({
			method: 'testMe3',
			params: args,
			expectReturn: false,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * 
	 */
	GeneratedTest.prototype.testMe4 = function() {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = 0;
		return this.rpc.call({
			method: 'testMe4',
			params: args,
			expectReturn: false,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * 
	 * @param {DefaultArray} p 
	 */
	GeneratedTest.prototype.TestDefaultArray = function(p) {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = Math.min(1, args.length);
		return this.rpc.call({
			method: 'TestDefaultArray',
			params: args,
			expectReturn: false,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * 
	 * @param {string} u 
	 * @returns {string}
	 */
	GeneratedTest.prototype.TestUrl = function(u) {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = Math.min(1, args.length);
		return this.rpc.call({
			method: 'TestUrl',
			params: args,
			expectReturn: true,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * 
	 * @returns {RenderOptions[]}
	 */
	GeneratedTest.prototype.getRenderOptions = function() {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = 0;
		return this.rpc.call({
			method: 'getRenderOptions',
			params: args,
			expectReturn: true,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * 
	 * @param {string} theString 
	 * @returns {Uint8Array}
	 */
	GeneratedTest.prototype.echoStringAsBuffer = function(theString) {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = Math.min(1, args.length);
		return this.rpc.call({
			method: 'echoStringAsBuffer',
			params: args,
			expectReturn: true,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * 
	 * @param {Uint8Array} buffer 
	 * @returns {number}
	 */
	GeneratedTest.prototype.getBufferSize = function(buffer) {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = Math.min(1, args.length);
		return this.rpc.call({
			method: 'getBufferSize',
			params: args,
			expectReturn: true,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * 
	 * @param {number} n 
	 * @returns {number[]}
	 */
	GeneratedTest.prototype.returnFrom0ToN = function(n) {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = Math.min(1, args.length);
		return this.rpc.call({
			method: 'returnFrom0ToN',
			params: args,
			expectReturn: true,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * 
	 * @param {boolean} required 
	 * @param {number} [p1] 
	 * @param {number} [p2] 
	 */
	GeneratedTest.prototype.optionalArgs = function(required, p1, p2) {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = Math.min(3, args.length);
		return this.rpc.call({
			method: 'optionalArgs',
			params: args,
			expectReturn: false,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * 
	 * @param {number[]} ints 
	 * @returns {number}
	 */
	GeneratedTest.prototype.sumArray = function(ints) {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = Math.min(1, args.length);
		return this.rpc.call({
			method: 'sumArray',
			params: args,
			expectReturn: true,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * 
	 * @param {object} a 
	 * @returns {object}
	 */
	GeneratedTest.prototype.testAny = function(a) {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = Math.min(1, args.length);
		return this.rpc.call({
			method: 'testAny',
			params: args,
			expectReturn: true,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * 
	 * @param {Date} timeParam 
	 * @returns {number}
	 */
	GeneratedTest.prototype.getSeconds = function(timeParam) {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = Math.min(1, args.length);
		return this.rpc.call({
			method: 'getSeconds',
			params: args,
			expectReturn: true,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * 
	 * @returns {Date}
	 */
	GeneratedTest.prototype.getNow = function() {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = 0;
		return this.rpc.call({
			method: 'getNow',
			params: args,
			expectReturn: true,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * 
	 * @returns {string}
	 */
	GeneratedTest.prototype.ns1.method1 = function() {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = 0;
		return this.rpc.call({
			method: 'ns1.method1',
			params: args,
			expectReturn: true,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * 
	 */
	GeneratedTest.prototype.ns1.sub1.sub2.method1 = function() {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = 0;
		return this.rpc.call({
			method: 'ns1.sub1.sub2.method1',
			params: args,
			expectReturn: false,
			transport: this.defaultTransport
		}, callback);
	};

	/**
	 * 
	 */
	GeneratedTest.prototype.ns2.sub1.sub2.method1 = function() {
		var args = Array.prototype.slice.call(arguments);
		var callback = null;
		if (args.length && typeof args[args.length - 1] === 'function') {
			callback = args.pop();
		}
		args.length = 0;
		return this.rpc.call({
			method: 'ns2.sub1.sub2.method1',
			params: args,
			expectReturn: false,
			transport: this.defaultTransport
		}, callback);
	};

}(typeof module !== 'undefined' ? module.exports : window));
